File: solutions/leetcode/56.合并区间.py
#
# @lc app=leetcode.cn id=56 lang=python3
#
# [56] 合并区间
#

# @lc code=start
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda inte: inte[0])
        ans = []
        l_hold, r_hold = intervals[0][0], intervals[0][1]
        for l_curr, r_curr in intervals:
            if l_curr > r_hold:
                ans.append([l_hold, r_hold])
                l_hold, r_hold = l_curr, r_curr
            else:
                r_hold = max(r_hold, r_curr)
        ans.append([l_hold, r_hold])
        return ans
        
# @lc code=end

File: solutions/leetcode/56.合并区间.cpp
/*
 * @lc app=leetcode.cn id=56 lang=cpp
 *
 * [56] 合并区间
 */

// @lc code=start
class Solution {
    public:
        static bool comp(vector<int>& intA, vector<int>& intB) {
            return intA[0] < intB[0];
        }
        vector<vector<int>> merge(vector<vector<int>>& intervals) {
            sort(intervals.begin(), intervals.end(), comp);
            vector<vector<int>> ans;
            int l = intervals[0][0], r = intervals[0][1];
            for (auto& interval : intervals) {
                if (interval[0] > r) {
                    ans.push_back({l, r});
                    l = interval[0];
                    r = interval[1];
                } else {
                    r = max(r, interval[1]);
                }
            }
            ans.push_back({l, r});
            return ans;
        }
    };
// @lc code=end

